plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt

```diff
@@ -137,9 +133,52 @@ class StmtConversionVisitor : FirVisitor<Exp, StmtConversionContext>() {
 
+        val leftType = data.embedType(equalityOperatorCall.arguments[0])
+        val rightType = data.embedType(equalityOperatorCall.arguments[1])
+
         return when (equalityOperatorCall.operation) {
-            FirOperation.EQ -> Exp.EqCmp(left, right)
-            FirOperation.NOT_EQ -> Exp.NeCmp(left, right)
+            FirOperation.EQ -> {
+                if (leftType is NullableTypeEmbedding && rightType !is NullableTypeEmbedding) {
+                    Exp.And(
+                        Exp.NeCmp(left, NullableDomain.nullVal(leftType.elementType.type)),
+                        // TODO: Replace the Eq comparison with a member call function to `left.equals`
+                        Exp.EqCmp(CastingDomain.cast(left, leftType.elementType.type), right.withType(left.type))
+                    )
+                } else if (leftType is NullableTypeEmbedding && rightType is NullableTypeEmbedding) {
+                    Exp.Or(
+                        Exp.And(
+                            Exp.EqCmp(left, NullableDomain.nullVal(leftType.elementType.type)),
+                            Exp.EqCmp(right, NullableDomain.nullVal(rightType.elementType.type)),
+                        ),
+                        // TODO: Replace the Eq comparison with a member call function to `left.equals`
+                        Exp.EqCmp(left, right.withType(left.type))
+                    )
+                } else {
+                    // TODO: Replace the Eq comparison with a member call function to `left.equals`
+                    Exp.EqCmp(left, right.withType(left.type))
+                }
+            }
+            FirOperation.NOT_EQ -> {
+                if (leftType is NullableTypeEmbedding && rightType !is NullableTypeEmbedding) {
+                    Exp.And(
+                        Exp.EqCmp(left, NullableDomain.nullVal(leftType.elementType.type)),
+                        // TODO: Replace the Ne comparison with a member call function to `left.equals`
+                        Exp.NeCmp(left.withType(leftType.elementType.type), right.withType(leftType.elementType.type))
+                    )
+                } else if (leftType is NullableTypeEmbedding && rightType is NullableTypeEmbedding) {
+                    Exp.Or(
+                        Exp.And(
+                            Exp.NeCmp(left, NullableDomain.nullVal(leftType.elementType.type)),
+                            Exp.NeCmp(right, NullableDomain.nullVal(rightType.elementType.type)),
+                        ),
+                        // TODO: Replace the Ne comparison with a member call function to `left.equals`
+                        Exp.NeCmp(left, right.withType(left.type))
+                    )
+                } else {
+                    // TODO: Replace the Ne comparison with a member call function to `left.equals`
+                    Exp.NeCmp(left, right.withType(left.type))
+                }
```

Comment:
Could this case be replaced with a `Not` wrapped around the `EQ` case?  Or perhaps both could be done the same way?

Reply:
Doing both the same way is probably not worth the extra logic as they are quite different. Converting it with `Not` is I think a good idea. We can remove the small additional complexity cost of `Not` later if the SMT solver time becomes a problem.